File: snippets/dart_snippets/lib/getting_started.dart
import 'dart:typed_data';

import 'package:breez_sdk/breez_sdk.dart';
import 'package:breez_sdk/bridge_generated.dart';

Future<void> initializeSDK() async {
  // ANCHOR: init-sdk

  // Initialize SDK logs listener
  BreezSDK().initialize();

  // Create the default config
  Uint8List seed = await BreezSDK().mnemonicToSeed("<mnemonic words>");
  String inviteCode = "<invite code>";
  String apiKey = "<api key>";
  NodeConfig nodeConfig = NodeConfig.greenlight(
    config: GreenlightNodeConfig(
      partnerCredentials: null,
      inviteCode: inviteCode,
    ),
  );
  Config config = await BreezSDK().defaultConfig(
    envType: EnvironmentType.Production,
    apiKey: apiKey,
    nodeConfig: nodeConfig,
  );

  // Customize the config object according to your needs
  config = config.copyWith(workingDir: "path to an existing directory");

  // Connect to the Breez SDK make it ready for use
  ConnectRequest connectRequest = ConnectRequest(config: config, seed: seed);
  return await BreezSDK().connect(req: connectRequest);
  // ANCHOR_END: init-sdk
}

Future<void> connectRestoreOnly(Config config, Uint8List seed) async {
  // ANCHOR: init-sdk-restore-only
  ConnectRequest connectRequest = ConnectRequest(config: config, seed: seed, restoreOnly: true);
  return await BreezSDK().connect(req: connectRequest);
  // ANCHOR_END: init-sdk-restore-only
}

Future<void> fetchBalance(String lspId) async {
  // ANCHOR: fetch-balance
  NodeState? nodeInfo = await BreezSDK().nodeInfo();
  if (nodeInfo != null) {
    int lnBalance = nodeInfo.channelsBalanceMsat;
    int onchainBalance = nodeInfo.onchainBalanceMsat;
    print(lnBalance);
    print(onchainBalance);
  }
  // ANCHOR_END: fetch-balance
}

// ANCHOR: logging
void onLogEntry(log) {
  print("Received log ${log.level}]: ${log.line}");
}

void logging(BreezSDK breezSDK) {
  breezSDK.logStream.listen(onLogEntry);
}
// ANCHOR_END: logging

File: snippets/swift/BreezSDKExamples/Sources/GettingStarted.swift
//
//  GettingStarted.swift
//
//
//  Created by ruben on 13/11/2023.
//

import BreezSDK

// ANCHOR: init-sdk
// SDK events listener
class SDKListener: EventListener {
    func onEvent(e: BreezEvent) {
        print("received event ", e)
    }
}

func gettingStarted() throws -> BlockingBreezServices? {
    // Create the default config
    let seed = try? mnemonicToSeed(phrase: "<mnemonic words>")

    let inviteCode = "<invite code>"
    let apiKey = "<api key>"
    var config = defaultConfig(envType: EnvironmentType.production, apiKey: apiKey,
                               nodeConfig: NodeConfig.greenlight(
                                   config: GreenlightNodeConfig(partnerCredentials: nil, inviteCode: inviteCode)))

    // Customize the config object according to your needs
    config.workingDir = "path to an existing directory"

    // Connect to the Breez SDK make it ready for use
    guard seed != nil else {
        return nil
    }
    let connectRequest = ConnectRequest(config: config, seed: seed!)
    let sdk = try? connect(req: connectRequest, listener: SDKListener())

    return sdk
}
// ANCHOR_END: init-sdk

func gettingStartedRestoreOnly(config: Config, seed: [UInt8]) throws -> BlockingBreezServices? {
    // ANCHOR: init-sdk-restore-only
    let connectRequest = ConnectRequest(config: config, seed: seed, restoreOnly: true)
    let sdk = try? connect(req: connectRequest, listener: SDKListener())
    // ANCHOR_END: init-sdk-restore-only
    return sdk
}

func gettingStartedNodeInfo(sdk: BlockingBreezServices) {
    // ANCHOR: fetch-balance
    if let nodeInfo = try? sdk.nodeInfo() {
        let lnBalance = nodeInfo.channelsBalanceMsat
        let onchainBalance = nodeInfo.onchainBalanceMsat
        print(lnBalance);
        print(onchainBalance);
    }
    // ANCHOR_END: fetch-balance
}

// ANCHOR: logging
class SDKLogStream: LogStream {
    func log(l: LogEntry) {
        print("Received log [", l.level, "]: ", l.line)
    }
}

func logging() throws {
    try? setLogStream(logStream: SDKLogStream())
}
// ANCHOR_END: logging

// ANCHOR: disconnect
func disconnect(sdk: BlockingBreezServices) throws {
    try? sdk.disconnect()
}
// ANCHOR_END: disconnect
